Add new section to lesson 4 (#149)
datacarpentry/r-intro-geospatial: episodes/04-data-structures-part2.Rmd

str(gapminder)
```

We can also examine individual columns of the data frame with the `class` or

> 'typeof' functions.:

Reviewer: ```suggestion
'typeof' functions:
```

I'd suggest just using `class()` here. In my experience, `typeof()` is a much less common function which has some non-intuitive behaviors, for instance returning `double` instead of `numeric` and returning `integer` for factors:

```r
> typeof(factor("a"))
[1] "integer"
```

This is because `typeof()` indicates the internal type (effectively, C type) of an object, rather than its R-level class. Since it's the R-level class that determines how everything in R will treat a variable, I think it makes more sense to emphasize `class()` alone here, to avoid confusing learners with multiple representations of the same concept.

Author: Would you like to open a new issue about your suggestion @mikemahoney218? I think it would be better that way.

To understand why R is giving us a warning when we try to add this row, let's learn a little more about factors.

## Removing columns and rows in data frames

To remove columns from a data frame, we can use the 'subset' function.

> This function allows us to remove columns using their names:

Reviewer 2: Maybe add a little context here? 
```suggestion
To remove columns from a data frame, we can use the 'subset' function.
This function allows us to remove columns using their names.
If we want to keep all columns except continent, pop and gdpPercap we can use the following `subset` command:
```

```{r}
life_expectancy <- subset(gapminder, select = -c(continent, pop, gdpPercap))
head(life_expectancy)
```

We can also use a logical vector to achieve the same result. Make sure the
vector's length match the number of columns in the data frame (to avoid vector

> recycling):

Reviewer 2: maybe a link to the definition of vector recycling or writing it in plain language here?

Author: @sstevens2  - I added a short description of what vector recycling is and a link to a book that explains more details. Please let me know if it is enough.

```{r}
life_expectancy <- gapminder[c(TRUE, TRUE, FALSE, FALSE, TRUE, FALSE)]
head(life_expectancy)
```

> Alternatively, we can use column's positions:

Reviewer 2: ```suggestion
Alternatively, we can use column positions:
```

```{r}
life_expectancy <- gapminder[-c(3, 4, 6)]
head(life_expectancy)
```

> Note that the easy way to remove rows from a data frame is selecting the rows

Reviewer 2: ```suggestion
Note that typically we select the rows we want to keep, rather than removing rows we do not want in the data.
```

we want to keep instead.

> Anyway, to remove rows from a data frame, we can use their positions:

Reviewer 2: ```suggestion
However, to remove rows from a data frame, we can use their positions:
```

```{r}
# Filter data for Afghanistan during the 20th century:
afghanistan_20c <- gapminder[gapminder$country == "Afghanistan" &
                             gapminder$year > 2000, ]

# Now remove data for 2002, that is, the first row:
afghanistan_20c[-1, ]

> ```

Reviewer 2: Feel like this section needs a bit more context on what data manipulation we are trying to do is before the code but I'm confused about the goal myself.  If we want to remove 2002 why not have the initial filter be > 2002?

Author: @sstevens2 - Can you please provide the context?

I think it is clear (because I've read this too many times!): The original dataset doesn't have any NAs, so, they add one and then remove it using `na.omit`. Basically they are showing how to use `na.omit`.

Perhaps we could add a syntetic example instead of messing with the original dataset? Would that be clearer?

Reviewer 2: I'll try to add something above.  I'm mostly confused about why were are subsetting to Afghanistan before we remove the row by position.

> An interesting case is removing rows containing NAs:

Reviewer 2: ```suggestion
In research, you may want to remove all the missing data prior to an analysis.  Let's first add some missing values (NAs) into the data and then we can use `na.omit()` to remove them.
```

```{r}
# Turn some values into NAs:
afghanistan_20c <- gapminder[gapminder$country == "Afghanistan", ]
afghanistan_20c[afghanistan_20c$year < 2007, "year"] <- NA

# Remove NAs
na.omit(afghanistan_20c)

> ```

Reviewer 2: may want to add some displays so they can see the NAs after they were added and see they were removed after.

Author: @sstevens2 - Do you mean to print the data frame with the NA rows?

That would work, but it could also detour too much from this episode's scope, which is exploring data frames.
I would argue that precisely because getting too much into details is that the reader, as you said, gets confused about the goal.

Reviewer 2: I'm realizing your `na.omit` code will print the data frame after the NA's are removed as written, but there isn't any code where learners actually see the NA we added in the data.  While you and I can read the line that adds the NA, it might not be intuitive what is happening to learners, they often rely on inspecting the data to see what has changed after each step.  I'll add a line as a suggestion above.

Author: @sstevens2 Besides, line 345 call `head(afghanistan_20c)` on the data frame and it shows some of the NAs added.

To understand why R is giving us a warning when we try to add this row, let's learn a little more about factors.

## Removing columns and rows in data frames

To remove columns from a data frame, we can use the 'subset' function.
This function allows us to remove columns using their names.
If we want to keep all columns except continent, pop and gdpPercap we can use the following `subset` command:

```{r}
life_expectancy <- subset(gapminder, select = -c(continent, pop, gdpPercap))
head(life_expectancy)
```

We can also use a logical vector to achieve the same result. Make sure the
vector's length match the number of columns in the data frame (to avoid vector

> recycling):

Reviewer 2: ```suggestion
vector's length match the number of columns in the data frame (to avoid R repeating the shorter vector to match the length of the longer vector):
```

Reviewer 2: tried to write a plain language version in line here.
Your vector recycling addition looks great!  It might be nice if it were closer to when you used the term and maybe it would be good in a spoiler fenced div. Alternatively maybe you could add the definition in glossario and then link it instead of having it in the text itself?

```{r}
life_expectancy <- gapminder[c(TRUE, TRUE, FALSE, FALSE, TRUE, FALSE)]
head(life_expectancy)
```

Vector recycling occurs when working with vectors of different length and it
consist on repeating the elements of the shorter vector up to the lenght of
the larger one. For more information, check the book R for Data Science and its
[chapter about vectors](https://r4ds.had.co.nz/vectors.html#scalars-and-recycling-rules).

Alternatively, we can use column positions:

```{r}
life_expectancy <- gapminder[-c(3, 4, 6)]
head(life_expectancy)
```

Note that typically we select the rows we want to keep, rather than removing rows we do not want in the data.

> we want to keep instead.

Reviewer 2: Hanging partial sentence here?
```suggestion
```

> However, to remove rows from a data frame, we can use their positions:

Reviewer 2: ```suggestion
However, to remove rows from a data frame, we can use their positions.
To practice on a smaller subset, we will filter the data to only those entries from Afghanistan after the year 2000.
This smaller dataset will be easier for us to inspect by eye and see the changes we are making.
```

```{r}
# Filter data for Afghanistan during the 20th century:
afghanistan_20c <- gapminder[gapminder$country == "Afghanistan" &
                             gapminder$year > 2000, ]

# Now remove data for 2002, that is, the first row:
afghanistan_20c[-1, ]
```

> In research, you may want to remove all the missing data prior to an analysis.  Let's first add some missing values (NAs) into the data and then we can use `na.omit()` to remove them.

Reviewer 2: ```suggestion
In research, we often remove rows based on features of the data itself, rather than its location.
For example, you may want to remove all the missing data prior to an analysis.  Let's first add some missing values (NAs) into the data and then we can use `na.omit()` to remove them.
```

```{r}
# Turn some values into NAs:
afghanistan_20c <- gapminder[gapminder$country == "Afghanistan", ]

> afghanistan_20c[afghanistan_20c$year < 2007, "year"] <- NA

Reviewer 2: ```suggestion
afghanistan_20c[afghanistan_20c$year < 2007, "year"] <- NA
head(afghanistan_20c)
```
didn't run it so hoping 2007 is in the top, if not we can add an argument here to see more.

Reviewer 2: or just include the object and see the whole object?

Author: @sstevens2 Yes, it is at the top (and everywhere) because `afghanistan_20c` only contains the afghanian rows (line 338) and the years before 2007 are NAs.

After the call to `na.omit()`, the only remainng row corresponds to 2007.

To understand why R is giving us a warning when we try to add this row, let's learn a little more about factors.

## Removing columns and rows in data frames

To remove columns from a data frame, we can use the 'subset' function.
This function allows us to remove columns using their names.
If we want to keep all columns except continent, pop and gdpPercap we can use the following `subset` command:

```{r}
life_expectancy <- subset(gapminder, select = -c(continent, pop, gdpPercap))
head(life_expectancy)
```

We can also use a logical vector to achieve the same result. Make sure the

> vector's length match the number of columns in the data frame (to avoid R repeating the shorter vector to match the length of the longer vector):

Reviewer 2: ```suggestion
vector's length match the number of columns in the data frame (to avoid R repeating the shorter vector to match the length of the longer vector, called "vector recycling"):
```

```{r}
life_expectancy <- gapminder[c(TRUE, TRUE, FALSE, FALSE, TRUE, FALSE)]
head(life_expectancy)
```

> Vector recycling occurs when working with vectors of different length and it
> consist on repeating the elements of the shorter vector up to the lenght of
> the larger one. For more information, check the book R for Data Science and its
> [chapter about vectors](https://r4ds.had.co.nz/vectors.html#scalars-and-recycling-rules).

Reviewer 2: ```suggestion
:::::: spoiler

### Vector Recycling

Vector recycling occurs when working with vectors of different length and it
consist on repeating the elements of the shorter vector up to the lenght of
the larger one. For more information, check the book R for Data Science and its
[chapter about vectors](https://r4ds.had.co.nz/vectors.html#scalars-and-recycling-rules).
::::::::
```
I added a mention of the term above to provide some context.  It is useful but I'd say extra, so added it as a spoiler here.

To understand why R is giving us a warning when we try to add this row, let's learn a little more about factors.

## Removing columns and rows in data frames

To remove columns from a data frame, we can use the 'subset' function.
This function allows us to remove columns using their names.
If we want to keep all columns except continent, pop and gdpPercap we can use the following `subset` command:

```{r}
life_expectancy <- subset(gapminder, select = -c(continent, pop, gdpPercap))
head(life_expectancy)
```

We can also use a logical vector to achieve the same result. Make sure the

> vector's length match the number of columns in the data frame (to avoid R repeating the shorter vector to match the length of the longer vector, called "vector recycling"):

Reviewer 3: Spelling mistake: vector's length match -> vector's length matches

Author: Done - Thanks!

```{r}
life_expectancy <- gapminder[c(TRUE, TRUE, FALSE, FALSE, TRUE, FALSE)]
head(life_expectancy)
```

:::::: spoiler

### Vector Recycling

Vector recycling occurs when working with vectors of different length and it

> consist on repeating the elements of the shorter vector up to the lenght of

Reviewer 3: spelling mistake: consist on -> consists of
up to the lenght -> up to the length

Author: Done - Thanks!
